client/network/sync/src/lib.rs

```diff
@@ -1421,22 +1472,56 @@ where
 			.map_err(|error: codec::Error| error.to_string())
 	}
 
-	fn poll(&mut self, cx: &mut std::task::Context) -> Poll<PollResult<B>> {
+	fn poll(
+		&mut self,
+		cx: &mut std::task::Context,
+	) -> Poll<PollBlockAnnounceValidation<B::Header>> {
 		while let Poll::Ready(Some(event)) = self.service_rx.poll_next_unpin(cx) {
 			match event {
 				ToServiceCommand::SetSyncForkRequest(peers, hash, number) => {
 					self.set_sync_fork_request(peers, &hash, number);
 				},
+				ToServiceCommand::RequestJustification(hash, number) =>
+					self.request_justification(&hash, number),
+				ToServiceCommand::ClearJustificationRequests => self.clear_justification_requests(),
+				ToServiceCommand::BlocksProcessed(imported, count, results) => {
+					for result in self.on_blocks_processed(imported, count, results) {
+						match result {
+							Ok((id, req)) => self.send_block_request(id, req),
+							Err(BadPeer(id, repu)) => {
+								self.network_service
+									.disconnect_peer(id, self.block_announce_protocol_name.clone());
+								self.network_service.report_peer(id, repu)
+							},
+						}
+					}
+				},
+				ToServiceCommand::JustificationImported(peer, hash, number, success) => {
+					self.on_justification_import(hash, number, success);
+					if !success {
+						info!(target: "sync", "💔 Invalid justification provided by {} for #{}", peer, hash);
+						self.network_service
+							.disconnect_peer(peer, self.block_announce_protocol_name.clone());
+						self.network_service.report_peer(
+							peer,
+							sc_peerset::ReputationChange::new_fatal("Invalid justification"),
+						);
+					}
+				},
 			}
 		}
 		self.process_outbound_requests();
 
 		if let Poll::Ready(result) = self.poll_pending_responses(cx) {
-			return Poll::Ready(PollResult::Import(result))
+			match result {
+				ImportResult::BlockImport(origin, blocks) => self.import_blocks(origin, blocks),
+				ImportResult::JustificationImport(who, hash, number, justifications) =>
+					self.import_justifications(who, hash, number, justifications),
+			}
```

Comment:
Because these messages are created in an object akin to `NetworkService` which basically just provides an interface for other protocols to interact with the syncing subsystem.

I addressed the other issue, now it's a loop.

Reply:
Where is the loop you are talking about? this `if let Poll::Ready` should be while let.